### PR TITLE
Switch to standard labels from Helm best practices

### PR DESCRIPTION
**Pull request by kingdonb**
https://helm.sh/docs/chart_best_practices/labels/#standard-labels

Suggesting this as a fix for #604

### CODE REVIEW

**Comment by kingdonb** [comment]
I have tested it out here and it's working for me:

https://github.com/kingdonb/csh-flux/blob/494960cfb1a78bad3334158fc26241e3f0c95174/clusters/home-workers/kured/kured-helmrelease.yaml#L12-L23

It is definitely a breaking change, during the upgrade I got the error that daemonsets could not have a change in labels as it was forbidden by their immutability rules. If this issue was already documented somewhere else then I missed it.

**Review by ckotzbauer — CHANGES_REQUESTED** [review]
Can you please bump the "version" field in the Chart.yaml to the next major version, add a short description to the chart-README how to migrate from 2.x to 3.x and ensure that the tests are green?
Thanks!

**Comment by kingdonb** [comment]
Absolutely.

I have just haphazardly deleted the `DaemonSet` to get around the issue, since it is not possible to change labels on the daemonset and labels are immutable.

I will go back and test whether the `helm upgrade --force` works with this, and likewise if Helm Controller users can expect using this to help or not:

```
$ k explain helmrelease.spec.upgrade.force
KIND:     HelmRelease
VERSION:  helm.toolkit.fluxcd.io/v2beta1

FIELD:    force <boolean>

DESCRIPTION:
     Force forces resource updates through a replacement strategy.
```

I am honestly not sure if this works, since I have read recently about how replacement strategy does not always honor the same semantics as delete-and-recreate. So we may actually have to advise the users to delete and then recreate their DaemonSets through a Helm upgrade following the deletion.

Either way, I will test it a few times and make sure what I'm adding in the documentation reads well and is correct. 👍 This may take a couple of iterations though, please bear with...

**Comment by kingdonb** [comment]
This is my PR but I suppose I'm going to encourage us to think about it for a little while before merging this, as a major version bump is no trifling matter, there may be other changes, and I can maintain this change in my own fork as long as I'm the only person who has needed it.

I would much rather bump the MAJOR chart version once we are certain there are no other breaking changes required, and so we can merge them all together at once – philosophically I think that it is better to bump the major version less in the entire lifetime of a project. I am not the maintainer on this project though, so if you feel stronger about this change than I do, feel free to merge and release. 👍

It will make my life easier not having to maintain a separate fork in the long run, but FWIW that's a sacrifice I'm more than willing to make for the moment

**Comment by kingdonb** [comment]
I think I've said everything I thought that needed to be explained in the `charts/kured/README.md` now 👍

**Comment by ckotzbauer** [comment]
Thanks for putting so much thought into the upgrade process and breaking change. I like that very much. :+1: 

A major version jump with big changes is not as bad or rare for helm charts as it is for the software itself. (Kured itself does not get a new major version).
Many charts get a major update every few months (or sometimes even every few weeks). These label-changes are special, because it is really one of the few situations where a complete reinstall is necessary. We need to make users aware of this, correct. That's why it has to be in the chart readme. (See https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack#upgrading-chart as an example, each major release is briefly described here). But since Kured does not have a state, a reinstallation is very easy. With a GitOps tool (e.g. Flux2), the existing HelmRelease has to be deleted first and then reinserted with a second commit with the new version.

**Comment by kingdonb** [comment]
I squashed all my changes here, I will add one more commit that highlights first that uninstall and reinstall will work.

Then I'm sure I can trim some of the text that followed. I'd like to keep some explanation about the immutable labels just so that folks know why this uninstall/reinstall is necessary, but we probably don't need the whole story about Helm Controller.

**Review by kingdonb — COMMENTED** [review]
👍 LGTM

File: charts/kured/README.md
Comment (ckotzbauer) [inline charts/kured/README.md:52]: (Z 46-52): This can be removed, as this is the Kubernetes behaviour which can't be changed. Maybe one sentence can be added to the previous text-block which shortly descibes, that changing labels on existing objects needs a recreation.

**Review by ckotzbauer — COMMENTED** [review]
Thanks for your efforts. I think this can be shortened on several places. This has much todo with default behaviours of Helm and Kubernetes (which should not be descibed in detail here) and the fact that each cluster operator has to know how he/she has to handle their infrastructure tools, which should also not be part of the Kured documentation.

File: charts/kured/README.md
Comment (kingdonb) [reply charts/kured/README.md:52]: It's not that changing labels on existing objects needs a recreation though, it's DaemonSet in particular which is immutable. (Anyone who uses Kured must use it in DaemonSet mode though, so the fact is probably not really necessary to mention.)

File: charts/kured/README.md
Comment (kingdonb) [reply charts/kured/README.md]: 853d476

File: charts/kured/README.md
Comment (ckotzbauer) [reply charts/kured/README.md:50]: Label-Selectors are immutable (e.g. in the PodSpec of the DaemonSet)

File: charts/kured/README.md
Comment (kingdonb) [reply charts/kured/README.md]: 14afbef890721ea135b3b128b3f10ce95ec46ace

File: charts/kured/README.md
Comment (kingdonb) [reply charts/kured/README.md]: e42317eca41bf8f273be3faa46f349218f399ac7

Comment (kingdonb) [reply charts/kured/README.md]: If there is a finer point in the language that I've gotten subtly wrong then it's beyond my understanding, welcome to suggest any corrections or further deletions if you prefer. I'll accept them.

File: charts/kured/README.md
Comment (kingdonb) [inline charts/kured/README.md:46]:
```suggestion
Label selectors on a DaemonSet are immutable so `helm upgrade --force` will fail too.
```

Comment (kingdonb) [reply charts/kured/README.md:46]: Is this the suggested change 👍

**Comment by kingdonb** [comment]
Noodling around with my releases, I think I see what you're saying, the concern about labels and labelselectors is not only for the daemonset; after my experimentation, I wound up with surprising combinations of labels in places I did not expect them and the end result is Prometheus selected my serviceMonitor, but the service failed to match!

Maybe better if the directions say more forcefully that only the uninstall and reinstall approach is supported.

File: charts/kured/README.md
Comment (ckotzbauer) [inline charts/kured/README.md:39]: Can you remove the "updateStrategy" here? It is unrelated to the labels.

**Review by ckotzbauer — COMMENTED** [review]
Thanks for your changes so far, one more comment.

**Comment by kingdonb** [comment]
I will squash the fixup commits once more, LMK if there's anything else you notice 👍

**Review by ckotzbauer — APPROVED** [review]

**Comment by ckotzbauer** [comment]
@jackfrancis @evrardjp @dholbach Does anybody want to take another look since it's a major chart change?

**Review by jackfrancis — APPROVED** [review]
lgtm

this will be a bit painful but there's nothing imminent _forcing_ users to upgrade to 3.x

**Comment by kingdonb** [comment]
The fortunate thing is that users should not need to do any configuration change, an uninstall-reinstall is totally sufficient. And also fortunate that there are no stateful components, so for any users who are doing GitOps, this reinstall should be very easy (as is pinning the chart version to `^2.0.0`.) 👍 

I'm not sure what that failure is, but I'm guessing it is a flaky test that can be ignored. It did not show up in any earlier builds.

**Pull request merged**
